wasm/src/lsp/parse.rs

```diff
@@ -0,0 +1,127 @@
+use std::collections::HashMap;
+use std::rc::Rc;
+use clvm_tools_rs::compiler::comptypes::{BodyForm, CompileErr, CompileForm};
+use clvm_tools_rs::compiler::sexp::SExp;
+use clvm_tools_rs::compiler::srcloc::Srcloc;
+
+use crate::lsp::types::{
+    IncludeData,
+    ParseScope,
+    ReparsedExp,
+    ReparsedHelper,
+    ScopeKind,
+};
+
+#[derive(Debug, Clone)]
+// A parsed document.
+pub struct ParsedDoc {
+    // Ignored means that it didn't look like chialisp language, possibly it's
+    // clvm or another sexp language that is semi-parsable as chialisp.
+    pub ignored: bool,
+    pub mod_kw: Option<Srcloc>,
+    // If we were able to run a frontend pass (even partially), compiled contains
+    // it.  CompileForm is the result of frontend and is used for analyzing
+    // chialisp in many different tools deriving from clvm_tools_rs.
+    pub compiled: CompileForm,
+    // The scope stack for this file.
+    pub scopes: ParseScope,
+    // Helpers in ReparsedHelper form.  We pulled these indiviually by identifying
+    // their ranges in the source.
+    pub helpers: HashMap<Vec<u8>, ReparsedHelper>,
+    // If present, the main expression in ReparsedExp form.
+    pub exp: Option<ReparsedExp>,
+    // Includes in the various files, indexed by included file.
+    pub includes: HashMap<Vec<u8>, IncludeData>,
+    // Index of hashed ranges (as Reparsed* data) to the names they bind.
+    pub hash_to_name: HashMap<Vec<u8>, Vec<u8>>,
+    // Chialisp frontend errors encountered while parsing.
+    pub errors: Vec<CompileErr>,
+}
+
+impl ParsedDoc {
+    pub fn new(startloc: Srcloc) -> Self {
```

Comment:
Because `startloc` here is a non-reference value,
it seems the argument is expected to be cloned like `let pd = ParsedDoc(loc.clone())`
But it ends up that it is cloned multiple times like below.
```rust
       let nil = SExp::Nil(startloc.clone());
       ParsedDoc {
            ...,
            compiled: CompileForm {
                loc: startloc.clone(),
                ...
            },
            scopes: ParseScope {
                region: startloc,
                ...
            },
            ...
        }
```
I think cloning `Srcloc` multiple times can be avoided by using a reference and it reduces memory overhead.
I know it's difficult to change to pass `startloc` as a reference because you need to update code in `clvm_tools_rs` to accept a reference of `Srcloc`,
but how about taking an intermediate step at first in `vscode-chialisp-lsp` and later apply it to `clvm_tools_rs`?

```rust
#[derive(Debug, Clone)]
pub struct ParsedDoc<'a> {
    pub mod_kw: Option<&'a Srcloc>,
    ...
}
...
impl<'a> ParsedDoc<'a> {
    pub fn new(startloc: &'a Srcloc) -> Self {
            ...,
            scopes: ParseScope {
                // Introducing additional `clone()` here temporarily
                // until `clvm_tools_rs` is updated to accept a reference of `Srcloc`
                region: startloc.clone(),
                ...,
            },
            ...
        }
    }
}
```

Reply:
Srcloc is meant to be light weight so it can be cloned relatively inexpensively.  It contains an Rc of the filename, a row and column and an option of another row and column.  Cloning it doesn't matter much and the lifetime management will be more complex than its worth IMO to not have ownership here.